dbms/src/Storages/Page/V3/BlobStore.cpp

```diff
@@ -509,7 +506,10 @@ void BlobStore::removePosFromStats(BlobFileId blob_id, BlobFileOffset offset, si
     const auto & stat = blob_stats.blobIdToStat(blob_id);
     {
         auto lock = stat->lock();
-        need_remove_stat = stat->removePosFromStat(offset, size, lock);
+        auto remaining_valid_size = stat->removePosFromStat(offset, size, lock);
+        // BlobFile which is read-only or with capacity larger than config.file_limit_size won't be reused for other write,
+        // so it's safe and necessary to remove it here.
+        need_remove_stat = ((stat->isReadOnly() || stat->file_total_caps > config.file_limit_size) && remaining_valid_size == 0);
```

Comment:
Why we need `stat->isReadOnly() || stat->file_total_caps > config.file_limit_size` here? Can it be `stat->isReadOnly()`?

Reply:
Fixed.